### PR TITLE
High level design PR: Add SDS dynamic secret

### PR DESCRIPTION
**Pull request by qiwzhang**
Signed-off-by: Wayne Zhang <qiwzhang@google.com>

This PR is not a real PR, it is a high level design proposal to support fetching dynamic secrets.

Main changes:
1) added SecretCallback interface to notify secret has been changed
2) added DynamicSecretProvider interface to fetch dynamic secret.
3) secretManager to hold static and dynamic secret separately.
4) use shared_ptr for Context objects so an new object can be created while old one is in-use.

Code Status:  
Passed compiling for files under /source/....

### CODE REVIEW

**Comment by qiwzhang** [comment]
@lizan @PiotrSikora Could you help to review it?

File: include/envoy/secret/dynamic_secret_provider.h [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+#include "envoy/secret/secret_callbacks.h"
+#include "envoy/ssl/tls_certificate_config.h"
+
+namespace Envoy {
+namespace Secret {
+
+/**
+ * An interface to fetch dynamic secret.
+ */
+class DynamicSecretProvider {
```
Comment (lizan) [inline include/envoy/secret/dynamic_secret_provider.h]: I'm curious why we need this and `createDynamicSecretProvider`, but not make SecretManager to manage callbacks.

File: source/common/secret/secret_manager_util.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+#include "common/common/fmt.h"
+#include "common/json/json_loader.h"
+#include "common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Secret {
+
+class SecretManagerUtil {
+public:
+  virtual ~SecretManagerUtil() {}
```
Comment (lizan) [inline source/common/secret/secret_manager_util.h]: not needed.

File: include/envoy/secret/secret_manager.h [hunk 3]
```diff
@@ -3,31 +3,46 @@
...
+  /**
+   * Add or update SDS config source. SecretManager starts downloading secrets from registered
+   * config source.
+   *
+   * @param sdsConfigSource a protobuf message object contains SDS config source.
+   * @param config_name a name that uniquely refers to the SDS config source.
+   * @return a hash string of normalized config source.
+   */
+  virtual DynamicSecretProviderSharedPtr
+  createDynamicSecretProvider(const envoy::api::v2::core::ConfigSource& config_source,
```
Comment (lizan) [inline include/envoy/secret/secret_manager.h]: from impl, this is not a pure create, but find one if there is existing, need a better name.

File: include/envoy/ssl/context.h [hunk 4]
```diff
@@ -32,12 +32,15 @@ class Context {
    */
   virtual std::string getCertChainInformation() const PURE;
 };
+typedef std::shared_ptr<Context> ContextSharedPtr;
 
 class ClientContext : public virtual Context {};
 typedef std::unique_ptr<ClientContext> ClientContextPtr;
+typedef std::shared_ptr<ClientContext> ClientContextSharedPtr;
```
Comment (lizan) [inline include/envoy/ssl/context.h]: Do we intended to make all *Context to SharedPtr? What is the usecase of unique_ptrs?

File: source/common/ssl/ssl_socket.h [hunk 5]
```diff
@@ -55,6 +56,7 @@ class SslSocket : public Network::TransportSocket,
   std::vector<std::string> getDnsSansFromCertificate(X509* cert);
 
   Network::TransportSocketCallbacks* callbacks_{};
+  ContextSharedPtr ctx_owner_;
```
Comment (lizan) [inline source/common/ssl/ssl_socket.h]: you should be able to have this as `std::shared_ptr<ContextImpl>` and cast it with `std::dynamic_pointer_cast`

**Comment by lizan** [comment]
This includes all changes in https://github.com/envoyproxy/envoy/pull/3700 right?

While we can review this on high level, but please address review/test on that first.

**Comment by JimmyCYJ** [comment]
@lizan Yes, I am modifying #3700 to make it consistent with this high level design. That PR currently needs more tests to meet test coverage requirements. Once the PR is ready, I will ping you.

File: include/envoy/ssl/context.h (hunk 4 above)
Comment (qiwzhang) [reply include/envoy/ssl/context.h]: unique_ptr can be removed.

Comment (qiwzhang) [reply include/envoy/ssl/context.h]: Actually, I am working on a separate PR of making ssl_socket not to hold context as its member, only use ctx at constructor.   If this works,  we will continue to use unique_ptr for ctx.

File: include/envoy/secret/secret_manager.h (hunk 3 above)
Comment (qiwzhang) [reply include/envoy/secret/secret_manager.h]: how about "findOrCreateDynamicSecretProvider"?

File: source/common/secret/secret_manager_util.h (hunk 2 above)
Comment (qiwzhang) [reply source/common/secret/secret_manager_util.h]: Will remove

File: source/common/ssl/ssl_socket.h (hunk 5 above)
Comment (qiwzhang) [reply source/common/ssl/ssl_socket.h]: As I mentioned above,  I try to remove this member variable ctx_.

**Comment by JimmyCYJ** [comment]
#3700 is under review, and we are updating this PR accordingly. Please keep this PR open.

**Pull request closed without merging**